File: app/src/main/java/com/crossbowffs/quotelock/modules/ModuleManager.java
package com.crossbowffs.quotelock.modules;

import android.content.Context;

import com.crossbowffs.quotelock.api.QuoteModule;
import com.crossbowffs.quotelock.modules.brainyquote.BrainyQuoteQuoteModule;
import com.crossbowffs.quotelock.modules.collections.CollectionsQuoteModule;
import com.crossbowffs.quotelock.modules.custom.CustomQuoteModule;
import com.crossbowffs.quotelock.modules.freakuotes.FreakuotesQuoteModule;
import com.crossbowffs.quotelock.modules.hitokoto.HitokotoQuoteModule;
import com.crossbowffs.quotelock.modules.jinrishici.JinrishiciQuoteModule;
import com.crossbowffs.quotelock.modules.natune.NatuneQuoteModule;
import com.crossbowffs.quotelock.modules.wikiquote.WikiquoteQuoteModule;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class ModuleManager {
    private static final Map<String, QuoteModule> sModules = new LinkedHashMap<>();

    static {
        addLocalModule(new HitokotoQuoteModule());
        addLocalModule(new WikiquoteQuoteModule());
        addLocalModule(new JinrishiciQuoteModule());
        addLocalModule(new FreakuotesQuoteModule());
        addLocalModule(new NatuneQuoteModule());
        addLocalModule(new BrainyQuoteQuoteModule());
        addLocalModule(new CustomQuoteModule());
        addLocalModule(new CollectionsQuoteModule());
    }

    private static void addLocalModule(QuoteModule module) {
        String className = module.getClass().getName();
        sModules.put(className, module);
    }

    public static QuoteModule getModule(Context context, String className) {
        QuoteModule module = sModules.get(className);
        if (module != null) {
            return module;
        }
        throw new ModuleNotFoundException("Module not found for class: " + className);
    }

    public static List<QuoteModule> getAllModules(Context context) {
        return new ArrayList<>(sModules.values());
    }
}

File: app/src/main/java/com/crossbowffs/quotelock/modules/ModuleManager.kt
package com.crossbowffs.quotelock.modules

import com.crossbowffs.quotelock.api.QuoteModule
import com.crossbowffs.quotelock.modules.brainyquote.BrainyQuoteQuoteModule
import com.crossbowffs.quotelock.modules.collections.CollectionsQuoteModule
import com.crossbowffs.quotelock.modules.custom.CustomQuoteModule
import com.crossbowffs.quotelock.modules.fortune.FortuneQuoteModule
import com.crossbowffs.quotelock.modules.freakuotes.FreakuotesQuoteModule
import com.crossbowffs.quotelock.modules.hitokoto.HitokotoQuoteModule
import com.crossbowffs.quotelock.modules.jinrishici.JinrishiciQuoteModule
import com.crossbowffs.quotelock.modules.libquotes.LibquotesQuoteModule
import com.crossbowffs.quotelock.modules.natune.NatuneQuoteModule
import com.crossbowffs.quotelock.modules.wikiquote.WikiquoteQuoteModule

object ModuleManager {
    private val sModules: MutableMap<String, QuoteModule> = LinkedHashMap()

    init {
        addLocalModule(HitokotoQuoteModule())
        addLocalModule(WikiquoteQuoteModule())
        addLocalModule(JinrishiciQuoteModule())
        addLocalModule(FreakuotesQuoteModule())
        addLocalModule(NatuneQuoteModule())
        addLocalModule(BrainyQuoteQuoteModule())
        addLocalModule(LibquotesQuoteModule())
        addLocalModule(FortuneQuoteModule())
        addLocalModule(CustomQuoteModule())
        addLocalModule(CollectionsQuoteModule())
    }

    private fun addLocalModule(module: QuoteModule) {
        val className = module::class.qualifiedName ?: ""
        sModules[className] = module
    }

    fun getModule(className: String): QuoteModule {
        val module = sModules[className]
        return module ?: throw ModuleNotFoundException("Module not found for class: $className")
    }

    fun getAllModules(): List<QuoteModule> {
        return ArrayList(sModules.values)
    }
}